src/jvmMain/kotlin/io/beatmaps/api/maps.kt

```diff
@@ -263,26 +265,29 @@ fun Route.mapDetailRoute() {
         }
     }
 
-    post<MapsApi.Validate> {
+    post<MapsApi.DeclareAi> {
         call.response.header("Access-Control-Allow-Origin", "*")
         requireAuthorization { user ->
-            if (!user.isAdmin()) {
-                call.respond(HttpStatusCode.BadRequest)
-            } else {
-                val mapUpdate = call.receive<ValidateMap>()
-
-                val result = transaction {
+            val mapUpdate = call.receive<AiDeclaration>()
+            val result = transaction {
+                if (!user.isAdmin() && !(isUploader(mapUpdate.id, user.userId) && mapUpdate.automapper)) {
+                    false
+                } else {
                     Beatmap.update({
                         (Beatmap.id eq mapUpdate.id)
                     }) {
-                        it[automapper] = mapUpdate.automapper
+                        it[declaredAi] = when {
+                            !mapUpdate.automapper && user.isAdmin() -> AiDeclarationType.None
```

Comment:
The `user.isAdmin()` check here isn't necessary as to be in this block it's either true or automapper isn't false

However I would probably rewrite this whole block as below to avoid doing a separate query to get the uploader:
```kotlin
val admin = user.isAdmin()
Beatmap.update({
    (Beatmap.id eq mapUpdate.id).let { q ->
        if (user.isAdmin()) {
            q // If current user is admin don't check the user
        } else {
            q and (Beatmap.uploader eq user.userId)
        }
    }
}) {
    it[declaredAi] = when {
        !mapUpdate.automapper && admin -> AiDeclarationType.None
        mapUpdate.automapper && admin -> AiDeclarationType.Admin
        else -> AiDeclarationType.Uploader
    }
    it[updatedAt] = NowExpression(updatedAt.columnType)
} > 0
```

The only situation this would have a different affect is when the uploader makes a call that tries to set automapper to false. Instead of doing nothing this would treat that as true which seems like a non-issue.